### PR TITLE
conway instances added

### PR DESCRIPTION
**Pull request by JaredCorduan**
# Description

I added instances for `ConwayDelegsPredFailure` which are required to integrate the latest ledger into consensus: `Typeable`, `NoThunks`, `EncCBOR`, `DecCBOR`, and `Arbitary`. The last three, however, are undefined.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the `CHANGELOG.md` for affected package
- [x] The version bounds in `.cabal` files are updated
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs]:
```suggestion
instance Typeable era => EncCBOR (ConwayDelegsPredFailure era) where
```

**Review by lehins — APPROVED** [review]
:+1:

**Comment by JaredCorduan** [comment]
@lehins I've removed all the `undefined`s since you reviewed. I'll wait for another :+1: from you before I merge.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs]: Every type in Haskell has `Typeable` instance, it does not need to be derived
```suggestion
  deriving Generic
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs]: Why not reduce clutter by embracing `LambdaCase`
```suggestion
  encCBOR = encode . \case
    DelegateeNotRegisteredDELEG kh -> Sum (DelegateeNotRegisteredDELEG @era) 0 !> To kh
    WithdrawalsNotInRewardsDELEGS rs -> Sum (WithdrawalsNotInRewardsDELEGS @era) 1 !> To rs
    CertFailure x -> Sum (CertFailure @era) 2 !> To x
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs]: This import will go too
```suggestion
```

File: eras/conway/impl/CHANGELOG.md
Comment (lehins) [inline eras/conway/impl/CHANGELOG.md]:
```suggestion
  `NoThunks`, `EncCBOR`, `DecCBOR`, and `Arbitrary`
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Delegs.hs:92]: FYI. Because these types are parametrized by `era`, it is possible to also easily add `FromCBOR` and `ToCBOR` as well. But, we can add them on the need to have basis

**Review by lehins — APPROVED** [review]
Looks great. Some minor suggestions

**Pull request merged**
